ipc/provider/src/checkpoint.rs

```diff
@@ -162,30 +178,62 @@ impl<T: BottomUpCheckpointRelayer + Send + Sync + 'static> BottomUpCheckpointMan
                     .await?;
                 log::debug!("bottom up bundle: {bundle:?}");
 
-                let epoch = self
-                    .parent_handler
-                    .submit_checkpoint(
-                        submitter,
-                        bundle.checkpoint,
-                        bundle.signatures,
-                        bundle.signatories,
-                    )
+                // We support parallel checkpoint submission using FIFO order with a limited parallelism (controlled by
+                // the size of submission_semaphore).
+                // We need to acquire a permit (from a limited permit pool) before submitting a checkpoint.
+                // We may wait here until a permit is available.
+                let parent_handler_clone = Arc::clone(&self.parent_handler);
+                let submission_permit = self
+                    .submission_semaphore
+                    .clone()
+                    .acquire_owned()
                     .await
-                    .map_err(|e| {
-                        anyhow!(
-                            "cannot submit bottom up checkpoint at height {} due to: {e:}",
-                            event.height
-                        )
-                    })?;
-
-                log::info!(
-                    "submitted bottom up checkpoint({}) in parent at height {}",
-                    event.height,
-                    epoch
+                    .unwrap();
+                all_submit_tasks.push(tokio::task::spawn(async move {
+                    Self::submit_checkpoint(parent_handler_clone, submitter, bundle, event).await;
```

Comment:
Will this actually work in parallel, because there is a `sequential` enforcement in the checkpoint height, so say we are submitting two checkpoints at height 100 and 130. If height 100 is not submitted, 130 cannot be submitted. My question is really about sort of `race` in the sense that if height 100 is still in the memory pool and the transaction for height 130 is submitted, will it pass the gas estimation without errors? Or will it be rejected due to `nonce` not set incrementally?

Reply:
I'm not entirely sure whether the submit of 130 can succeed if the submit of 100 is still in progress...@aakoshh Can you share your opinions here?

Reply:
I don't know how Lotus works exactly, but let's say that it:
* limits the number of pending transactions in the mempool to 4
* orders the pending transactions by nonce 
* applies them to an in-memory check-state like fendermint

If that were true, then as long as the transactions are _created_ in a logical order, their order could be restored by Lotus when they are included in the block. As for the checks, it probably depends on whether that particular node has seen the preceding transactions are not. 

Note that we would still _send_ the transactions sequentially, we just don't want to _wait_ for the receipt between each submission.

Reply:
Indeed, blockchain nodes accept gapped transactions via their RPC and via the mpool, because they live in an eventually consistent universe. Note that any of these can happen, in addition to other scenarios, of course:

* The user submits transactions in sequence to a load-balanced RPC endpoint, and they land gapped in the backend nodes.
* The user submits out of order transactions to the same node, which is just a subcase of the gapped situation.
* The user submits many transactions, all in order, to the same backend node, but each tx takes different gossip propagation routes through the network and it arrives to various nodes in various non-sequential and gapped orders.